src/infinitempomatrix.jl

```diff
@@ -44,111 +44,191 @@ function InfiniteMPOMatrix(data::Vector{Matrix{ITensor}}, translator::Function)
+    end
+  else
+    right_ind = only(uniqueinds(tensor, sit, left_ind))
+  end
+  left_dim = dim(left_ind)
+  right_dim = dim(right_ind)
+  #Build the local projectors.
+  left_tags = get(kwargs, :left_tags, tags(left_ind))
+  top_left, middle_left, bottom_left = build_three_projectors_from_index(
+    left_ind; tags=left_tags
+  )
+  right_tags = get(kwargs, :righ_tags, tags(right_ind))
+  top_right, middle_right, bottom_right = build_three_projectors_from_index(
+    right_ind; tags=right_tags
+  )
+
+  matrix = fill(op("Zero", local_sit), 3, 3)
+  for (idx_left, proj_left) in enumerate([top_left, middle_left, bottom_left])
+    for (idx_right, proj_right) in enumerate([top_right, middle_right, bottom_right])
+      matrix[idx_left, idx_right] = proj_left * tensor * proj_right
     end
-    push!(inds_array, inds_n)
   end
-  return inds_array
+  return matrix
+end
+
+function convert_itensor_3vector(
+  tensor; leftdir=ITensors.In, first=false, last=false, kwargs...
+)
+  @assert order(tensor) == 3
+  #Identify the different indices
+  sit = filterinds(inds(tensor); tags="Site")
+  local_sit = dag(only(filterinds(sit; plev=0)))
+  #A bit roundabout as filterinds does not accept dir
+  old_ind = only(uniqueinds(tensor, sit))
+  if dir(old_ind) == leftdir || last
+    new_tags = get(kwargs, :left_tags, tags(old_ind))
+    top, middle, bottom = build_three_projectors_from_index(old_ind; tags=new_tags)
+    vector = fill(op("Zero", local_sit), 3, 1)
+  else
+    new_tags = get(kwargs, :right_tags, tags(old_ind))
+    top, middle, bottom = build_three_projectors_from_index(old_ind; tags=new_tags)
+    vector = fill(op("Zero", local_sit), 1, 3)
+  end
+  for (idx, proj) in enumerate([top, middle, bottom])
+    vector[idx] = proj * tensor
+  end
+  return vector
 end
 
-#
-#  Hm is the InfiniteMPOMatrix
-#  Hlrs is an array of {ITensor,Index,Index}s, one for each site in the unit cell.
-#  Hi is a CelledVector of ITensors.
-#
-function InfiniteMPO(Hm::InfiniteMPOMatrix)
-  inds_array = find_all_links(Hm)
-  Hlrs = cat_to_itensor.(Hm, inds_array) #return an array of {ITensor,Index,Index}
-  #
-  #  Unpack the array of tuples into three arrays.  And also get an array site indices.
-  #
-  Hi = CelledVector([Hlr[1] for Hlr in Hlrs], translator(Hm))
-  ils = CelledVector([Hlr[2] for Hlr in Hlrs], translator(Hm))
-  irs = CelledVector([Hlr[3] for Hlr in Hlrs], translator(Hm))
-  site_inds = [commoninds(Hm[j][1, 1], Hm[j][end, end])[1] for j in 1:nsites(Hm)]
-  #
-  #  Create new tags with proper cell and link numbers.  Also daisy chain
-  #  all the indices so right index at j = dag(left index at j+1)
-  #
-  for j in 1:nsites(Hm)
-    newTag = "Link,c=$(getcell(site_inds[j])),l=$(getsite(site_inds[j]))"
-    ir = replacetags(irs[j], tags(irs[j]), newTag) #new right index
-    Hi[j] = replaceinds(Hi[j], [irs[j]], [ir])
-    Hi[j + 1] = replaceinds(Hi[j + 1], [ils[j + 1]], [dag(ir)])
+function apply_tensor(A::Array{ITensor,N}, B::ITensor...) where {N}
```

Comment:
This doesn't really seem like it needs to be a function. For example it could just be done with:
```julia
map(a -> contract(a, B...), A)
```
if I understand it correctly.

Reply:
Or `contract.(A, Ref.(B)...)` should also work.

Reply:
Removed the function and replaced it in the test.
It was convenient though not necessary indeed.